### DISCUSSION TITLE
onClose hook with database IO operations

### DISCUSSION THREAD

**Discussion by ghost** [fastify/fastify#4964] [Q&A] [answered]
Hi, I was wondering if there is some clean way to use the `onClose` hook in order to write on database some information before the fastify instance get torn down. In particular I'm using `fastify-mongodb` in order to connect to my MongoDB instance and I have a plugin that registers a `onClose` hook that updates the database before the server gets closed. I suspect that, since there is no way to specify an order among `onClose` hooks, the `fastify-mongodb` gets closed (and its client too) before the other `onClose` hook gets called. In this case I probably have a race condition of some sort so my plugin tries to connect to Mongo after the mongo client is trying to be closed, so the shutdown fails and gets stuck (without tangible errors). I use `close-with-grace` in order to gracefully shut down my fastify instance. I even use an older fastify version (3.9.2) that I cannot upgrade in the short term (production server) and so I cannot even try to use the newer `preClose` hook (if it can actually be useful). Any suggestions?

**Comment by mcollina** [comment, marked as answer]
Fastify v3 it out of LTS, upgrade.

**Reply by ghost** [reply]
I know 😢
